datafusion/functions/src/utils.rs

```diff
@@ -140,38 +140,25 @@ where
     F: Fn(L::Native, R::Native) -> Result<O::Native, ArrowError>,
     R::Native: TryFrom<ScalarValue>,
 {
-    Ok(match right {
+    let left = left.as_primitive::<L>();
+    let right = right.cast_to(&R::DATA_TYPE, None)?;
```

Comment:
This right cast wasn't applied if it was a scalar, so cast the columnarvalue itself to align both arms